### PR TITLE
fix: table index error issue (#1919)

### PR DESCRIPTION
**Pull request by msidolphin**
Please make sure these boxes are checked before submitting your PR, thank you!

bugfix #1919 #2247 

* [x] Make sure you follow Element's contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
* [x] Make sure you are merging your commits to `dev` branch.
* [x] Add some descriptions and refer to relative issues for your PR.

### CODE REVIEW

**Comment by iamkun** [comment]
@justwiner Can you review this PR, pls?

File: packages/table/__tests__/table.spec.ts [hunk 1]
```diff
@@ -994,69 +994,81 @@ describe('Table.vue', () => {
...
+          }
+        },
+      })
+      await nextTick()
+      const rows = wrapper.findAll('.el-table__row')
+      rows.forEach((row, index) => {
+        const cell = row.find('td')
+        expect(cell.text()).toMatch(`${index + 1}`)
+      })
+      !(async () => {
```
Comment (HerringtonDarkholme) [inline packages/table/__tests__/table.spec.ts:1059]: Why is this wrapped in an async iife?

File: packages/table/src/table-body/render-helper.ts [hunk 2]
```diff
@@ -145,7 +145,7 @@ function useRender<T>(props: Partial<TableBodyProps<T>>) {
        * TIP: One dimensional array is used temporarily to avoid rendering flicker.
        * The case of $index being modified has not been found by testing
```
Comment (HerringtonDarkholme) [inline packages/table/src/table-body/render-helper.ts:146]: The comment is outdated, can you update the comment here?

**Review by HerringtonDarkholme — CHANGES_REQUESTED** [review]
Some minor issues need update

File: packages/table/__tests__/table.spec.ts (hunk 1 above)
Comment (msidolphin) [reply packages/table/__tests__/table.spec.ts]: sorry, i made a mistake

File: packages/table/src/table-body/render-helper.ts (hunk 2 above)
Comment (msidolphin) [reply packages/table/src/table-body/render-helper.ts]: of course

**Review by HerringtonDarkholme — APPROVED** [review]

**Pull request merged**